### PR TITLE
Add a UI for displaying vector layer features in a grid

### PR DESCRIPTION
**Pull request by schmidtk**
This adds a few new directives for displaying vector data in a grid, along with some basic interaction via the list's context menu.

- `sourcegrid` is the base grid UI backed by Slickgrid.
- `featurelist` is a window to display features for a source, and can be launched from the Layers menu.

This PR also deprecates `datagrid`, which has not been supported for some time and will be removed in the future.

### CODE REVIEW

File: src/os/ui/grid/datagrid.js [hunk 1]
```diff
@@ -26,6 +31,232 @@ os.ui.grid.dataGridDirective = function() {
...
+
+
+/**
+ * @param {!angular.Scope} $scope
+ * @param {!angular.JQLite} $element
+ * @param {!angular.$timeout} $timeout
+ * @constructor
+ * @ngInject
+ *
+ * @deprecated Please use `os.ui.slick.SlickGridCtrl` instead.
```
[the remark this answers is not in the document]
Comment (schmidtk) [reply src/os/ui/grid/datagrid.js]: I consolidated the files so it's all in one place, but we may still remove `datagrid` in this PR. It hasn't been maintained/supported for quite some time.

**Comment by schmidtk** [comment]
@bradh I loaded that KML and I'm not seeing the duplicate `description`/`DESCRIPTION` columns in your screenshot. I only see the lower-case variant, and the Show text displays the description in both cases. Is the example you posted exactly what you're loading?

**Comment by bradh** [comment]
That KML is the first "row" of https://www.data.brisbane.qld.gov.au/data/dataset/c31cdaa4-2407-4aee-9322-92b95032ef59/resource/750d5fe4-fdeb-495c-aa8f-90ec81beb838/download/qppairportols.kml

Will try to upload it.

**Comment by bradh** [comment]
[qppairportols.kml.txt](https://github.com/ngageoint/opensphere/files/3195756/qppairportols.kml.txt)

**Comment by bradh** [comment]
Added a PR against your branch for the easy changes,

**Comment by schmidtk** [comment]
The description issue was caused by features having two description fields, `description` from the KML tag and `DESCRIPTION` from the `ExtendedData`. The description column formatter was not reading the proper field from the column definition, and instead just trying upper/lower case "description". That has been fixed, so the Show links in the grid should work better now. Feature Info uses different logic, so those show links still toggle the Description tab with the KML tag content. That felt outside the scope of this PR to investigate further.

I switched the icon, but took a different approach for the label. The menu option is now under the Layer group after Show Description, and has been named Show Features (with an updated tooltip). Different users resonate with list/table/grid/etc and none of those felt necessary in the menu label.

File: src/os/metrics/layersmetrics.js [hunk 2]
```diff
@@ -328,8 +328,8 @@ os.metrics.LayersMetrics = function() {
     key: os.metrics.Layer.EXPORT
   });
   this.addChild(contextFeatureLayersMenuLeaf, {
-    label: 'Feature List',
-    description: 'Display a list of features in the layer.',
+    label: 'Show Features',
+    description: 'Displays metadata for features in the layer.',
```
Comment (bradh) [inline src/os/metrics/layersmetrics.js]: To me its attributes, not metadata.

File: src/os/ui/featurelist.js [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    os.ui.window.bringToFront(windowId);
+  } else {
+    var scopeOptions = {
+      'source': source
+    };
+
+    var windowOptions = {
+      'id': windowId,
+      'label': 'Feature List: ' + source.getTitle(true),
+      'icon': 'fa fa-list-alt',
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/featurelist.js]: Resolved.

File: src/os/metrics/layersmetrics.js [hunk 4]
```diff
@@ -327,6 +327,11 @@ os.metrics.LayersMetrics = function() {
     description: 'Exports data from this layer.',
     key: os.metrics.Layer.EXPORT
   });
+  this.addChild(contextFeatureLayersMenuLeaf, {
+    label: 'Feature List',
+    description: 'Display a list of features in the layer.',
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/metrics/layersmetrics.js]: Resolved

File: src/os/metrics/layersmetrics.js [hunk 5]
```diff
@@ -327,6 +327,11 @@ os.metrics.LayersMetrics = function() {
     description: 'Exports data from this layer.',
     key: os.metrics.Layer.EXPORT
   });
+  this.addChild(contextFeatureLayersMenuLeaf, {
+    label: 'Feature List',
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/metrics/layersmetrics.js]: Resolved

File: src/os/ui/menu/layermenu.js [hunk 6]
```diff
@@ -205,6 +206,15 @@ os.ui.menu.layer.setup = function() {
         handler: os.ui.menu.layer.onExport_,
         metricKey: os.metrics.Layer.EXPORT,
         sort: -1000 // commonly used, so prioritize above other items
+      },
+      {
+        label: 'Feature List',
+        eventType: os.action.EventType.FEATURE_LIST,
+        tooltip: 'Display a list of features in the layer',
+        icons: ['<i class="fa fa-fw fa-list-alt"></i>'],
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/layermenu.js]: Resolved

File: src/os/ui/menu/layermenu.js [hunk 7]
```diff
@@ -205,6 +206,15 @@ os.ui.menu.layer.setup = function() {
         handler: os.ui.menu.layer.onExport_,
         metricKey: os.metrics.Layer.EXPORT,
         sort: -1000 // commonly used, so prioritize above other items
+      },
+      {
+        label: 'Feature List',
+        eventType: os.action.EventType.FEATURE_LIST,
+        tooltip: 'Display a list of features in the layer',
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/layermenu.js]: Resolved

File: src/os/ui/menu/layermenu.js [hunk 8]
```diff
@@ -205,6 +206,15 @@ os.ui.menu.layer.setup = function() {
         handler: os.ui.menu.layer.onExport_,
         metricKey: os.metrics.Layer.EXPORT,
         sort: -1000 // commonly used, so prioritize above other items
+      },
+      {
+        label: 'Feature List',
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/layermenu.js]: Resolved.

File: src/os/metrics/layersmetrics.js (hunk 2 above)
Comment (schmidtk) [reply src/os/metrics/layersmetrics.js]: This is one of those things that will never resonate with everyone, kind of like the term "features" itself. It varies between users, and developers tend to have different opinions yet. I'm not sure what the best approach is, other than removing the "metadata for" phrase entirely.

**Comment by schmidtk** [comment]
I fully agree with the sentiment on when to show certain menu items, but we may need to revisit the performance of determining those things. `os.source.Vector#getSelectedItems` returns a copy of the array, `getHiddenItems` is worse because it creates a feature array from the RBush and filters from there. While those are all fine under small loads, they will be a problem when working with large volumes of features. I'll defer that work for now to avoid additional scope creep in this PR.

File: src/os/ui/menu/featuremenu.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    handler: opt_handler,
+    sort: 1
+  });
+  group.addChild({
+    label: 'Display All',
+    eventType: prefix + os.action.EventType.DISPLAY_ALL,
+    tooltip: 'Displays all items',
+    icons: ['<i class="fa fa-fw fa-eye"></i>'],
+    handler: opt_handler,
+    sort: 2
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/featuremenu.js]: Resolved.

File: src/os/ui/menu/featuremenu.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    handler: opt_handler,
+    sort: 0
+  });
+  group.addChild({
+    label: 'Hide Unselected',
+    eventType: prefix + os.action.EventType.HIDE_UNSELECTED,
+    tooltip: 'Hides unselected items',
+    icons: ['<i class="fa fa-fw fa-eye-slash"></i>'],
+    handler: opt_handler,
+    sort: 1
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/featuremenu.js]: Resolved.

File: src/os/ui/menu/featuremenu.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    type: os.ui.menu.MenuItemType.GROUP,
+    sort: 1
+  });
+
+  group.addChild({
+    label: 'Hide Selected',
+    eventType: prefix + os.action.EventType.HIDE_SELECTED,
+    tooltip: 'Hides selected items',
+    icons: ['<i class="fa fa-fw fa-eye-slash"></i>'],
+    handler: opt_handler,
```
[the remark this answers is not in the document]
Comment (bradh) [reply src/os/ui/menu/featuremenu.js]: Resolved.

File: src/os/metrics/layersmetrics.js (hunk 2 above)
Comment (bradh) [reply src/os/metrics/layersmetrics.js]: I think leaving the `metadata for` part out would be better. Its a very minor nit though.

Comment (schmidtk) [reply src/os/metrics/layersmetrics.js]: I agree, removed.

**Comment by justin-bits** [comment]
https://github.com/ngageoint/opensphere/pull/567

**Pull request merged**